# Project: UPI Fraud Detector

# 1. Dataset Audit

In [1]:
import pandas as pd
import numpy as np

## 1.1 Load Dataset

In [2]:
df = pd.read_csv("../data/raw/fraud_dataset.csv")
df.shape

(26393, 65)

## 1.2 Column Overview

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26393 entries, 0 to 26392
Data columns (total 65 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   transaction_id                           26393 non-null  object 
 1   user_id                                  26393 non-null  object 
 2   merchant_id                              26393 non-null  object 
 3   amount                                   26393 non-null  float64
 4   timestamp                                26393 non-null  object 
 5   description                              26393 non-null  object 
 6   device_id                                26393 non-null  object 
 7   ip_address                               26393 non-null  object 
 8   location                                 26393 non-null  object 
 9   session_duration                         26393 non-null  int64  
 10  authentication_attempts                  26393

## 1.3 Target Variable Distribution

In [4]:
df['is_fraud'].value_counts()
df['is_fraud'].value_counts(normalize=True) * 100

is_fraud
0    82.779525
1    17.220475
Name: proportion, dtype: float64

## 1.4 Missing Values Check

In [5]:
df.isnull().sum().sort_values(ascending=False).head(10)

request_description    25661
url_referrer           25636
transaction_id             0
amount                     0
user_id                    0
description                0
device_id                  0
ip_address                 0
merchant_id                0
session_duration           0
dtype: int64

## 1.5 Duplicate Check

In [6]:
df.duplicated().sum()

np.int64(0)

## 1.6 ID Columns Uniqueness Check

In [7]:
print("Total rows:", len(df))
print("Unique transaction_id:", df['transaction_id'].nunique())
print("Unique user_id:", df['user_id'].nunique())
print("Unique device_id:", df['device_id'].nunique())
print("Unique merchant_id:", df['merchant_id'].nunique())

Total rows: 26393
Unique transaction_id: 26393
Unique user_id: 4818
Unique device_id: 6455
Unique merchant_id: 6544


## 1.7 Timestamp Range Check

In [10]:
df['timestamp'].head(10)

0    18:27.7
1    34:55.5
2    49:42.8
3    16:29.9
4    39:44.7
5    02:41.8
6    17:02.2
7    07:36.1
8    13:49.7
9    00:36.2
Name: timestamp, dtype: object

Note: timestamp column is in MM:SS.s format, not a valid datetime. Using `transaction_time_of_day` instead for hour-based features.

In [11]:
df['transaction_time_of_day'].describe()

count    26393.000000
mean        11.007426
std          7.071321
min          0.000000
25%          5.000000
50%         10.000000
75%         17.000000
max         23.000000
Name: transaction_time_of_day, dtype: float64

## 1.8 Audit Summary

- Shape: 26393 rows, 65 columns
- Target: is_fraud (17.22% fraud, 82.78% genuine)
- Missing values: url_referrer (97.1%), request_description (97.2%) — to be dropped
- Duplicates: 0
- user_id repeats (4818 unique) → velocity/frequency features feasible
- timestamp column is corrupted (MM:SS.s format) — not usable as datetime; use transaction_time_of_day instead
- Dataset appears synthetically generated (behavioral biometric columns unrealistic for real UPI data)